### PR TITLE
Average shared span for multiple matches

### PR DESCRIPTION
**Pull request by hfr1tz3**
Option 2 for issue #11 

For each node $n_2\in T_2$ with multiple matches in $T_1$, Let $\beta(n_2)=[n_1\in T_1 \colon \alpha(n_1)=n_2]$. Then we compute the similarity between two trees $T_1$, and $T_2$ as
$$sim(T_1,T_2)=\sum_{n_2\in T_2}\frac{1}{\beta(n_2)}\sum_{n_1\in \alpha^{-1}(n_2)} m(n_1,n_2),$$
which is the average over shared spans between nodes in $T_2$ and their multiple matches $\beta(n_2)$.

### CODE REVIEW

**Comment by hfr1tz3** [comment]
I added extra tests that are for the `ties='average'` method. I am open to more suggestions if we need more.

**Comment by petrelharp** [comment]
Gee, that's a quite clean implementation. Nice.

My first inclination is not to introduce the "ties" argument? I think we don't have a use case for `ties=None`, so why complicate things?

**Comment by hfr1tz3** [comment]
That's a fair point. I think it was sort of an idea to put in place, where in case we want to add a new matching scheme then we have the scaffolding for that already. However, if want to remove that I can. 
@petrelharp just ping me if you want it removed. I'll try to get to it while at JMM.

**Comment by hfr1tz3** [comment]
Alright, I removed the `ties` argument from the compare function. I additionally added another test case to check that the new average node span is working properly.

**Comment by petrelharp** [comment]
Hm, okay - sorry I didn't pick up on this earlier - but, this is also changing *dis*similarity (since both dissimilarity and tpr depend on `total_match_span`). Now I need to think about whether this is desired or not.

**Comment by petrelharp** [comment]
Okay: I think that if T1 is inferred and T2 is truth, then our motivation has been:
1. we want `ARF` to be "how much of T1 is wrong" (so, denominator is `|T1|`)
2. and `TPR` to be "how much of T2 is correctly represented in T1" (denominator is `|T2|`)
3. previously, interpretation (2) does not hold, because adding more nodes to T1 that mapped to the same node in T2 would change TPR when it wasn't actually changing how much of T2 that was represented
4. with this change, interpretation (1) arguably doesn't hold, because suppose those multiple nodes in T1 that map to a single node in T2 all agree with the node in T2; they are arguably not *wrong*, so should not count against it
5. the question is, I guess, what is "wrong"; one interpretation is that a node in T1 is 'wrong' if it implies relationships *not* in T2 (e.g., "on this chunk of genome A and B are more closely related than to C")
6. having multiple nodes in T1 that match a given node in T2 feels a bit wrong, but counterpoint: adding completely-unary nodes to T1 should arguably not be penalised because we know that those nodes in fact exist - edges are but chains of inheritance through a bunch of individuals

So: currently we have:
Screenshot from 2025-01-16 05-53-34
 ... so I guess I'm arguing that the right-hand equality in equation (2) above should not hold, nor the right hand equality here:
 Screenshot from 2025-01-16 05-56-17
so that, conceptually, `sim( )` is a sum over nodes in T2 and `dis( )` is a sum over nodes in T1. I think this argues that we need different names that sound less symmetric for them, unfortunately.

What do you think?

**Comment by hfr1tz3** [comment]
> ...  conceptually, `sim( )` is a sum over nodes in T2 and `dis( )` is a sum over nodes in T1. I think this argues that we need different names that sound less symmetric for them, unfortunately.

I actually like this change. When I was attempting to redefine terms in the paper, the notation got bogged down.

ARF in my mind should be defined as it was with

$$\textnormal{ARF}(T_1, T_2) = 1-\frac{1}{|| T_1||} \sum_{n\in N_1} m(n_1, \alpha(n_1)).$$

To comment about points 3, 4, and 6:
I took some time to think about possibly doing a 'weighted avg' for TPR if we have multiple matches for a node in $T_2$, however those still can have a TPR of over 1. I think with the averaged node span for TPR, we can drastically decrease TPR when that doesn't reflect the actual property of "how much of $T_2$ is in $T_1$". For instance, I feel like a tree sequence $T_1$ with extra unary nodes should have TPR of 1 when compared to its simplified version $T_2$.

If we want to remove the 'averaged node span' from TPR we could alternatively define it as

$$ \textnormal{TPR}(T_1, T_2) = \frac{1}{||T_2||}\sum_{n_2\in N_2} \max_{n_1\in \alpha^{-1}(n)}m(n_1, n_2)$$

Below is an example between avg TPR and max TPR:

image

**Comment by petrelharp** [comment]
Hey, this is a great point. `max` is totally better than `mean` here. Totally on board with this.

**Comment by hfr1tz3** [comment]
@petrelharp I made the change. 

The work is pretty much that there are "best matches" with respect to $T_1$ and $T_2$ resp (marked by n1 or n2). If its an n1 match we use that to compute dissimilarity and ARF and if its an n2 match that we use that to compute the "inverse dissimilarity" := total span $T_2$ - best_n2_match_span and tpr. 

I tried to make sure I added all the correct docs and tests, but let me know if I missed something.

**Comment by petrelharp** [comment]
This looks right, although my brain's a bit slow right now. I'm not sure about the name "inverse dissimilarity" (but, good idea); for clarification: is `compare(t1, t2).inverse_dissimilarity == compare(t2, t1).dissimilarity`? (No, right? What's the difference?)

And, I've invited you as a collaborator so your tests run automatically (thought we'd done that already?). See the errors there, e.g.
```
FAILED tests/test_methods.py::TestDissimilarity::test_basic_comparison[pair0] - TypeError: ARFResult.__init__() got an unexpected keyword argument 'inverse_dissimilarity'
```

File: tscompare/methods.py [hunk 1]
```diff
@@ -348,13 +352,23 @@ def compare(ts, other, transform=None):
     If there are multiple matches with the same longest shared span
     for a single node, the best match is the match that is closest in time.
 
+    For each node in `other` we compute the best matched span
+    as the average shared span amongst all nodes in `ts` which are its match.
```
Comment (petrelharp) [inline tscompare/methods.py:356]: whoops - this shouldn't say "average" any more, right?

**Comment by petrelharp** [comment]
And, maybe we (sigh) need another term, since we use "best match of n1 is n2" to mean something different than "best match of n2 is n1" (since "best match": T1 -> T2 is many-to-one).

**Comment by hfr1tz3** [comment]
> This looks right, although my brain's a bit slow right now. I'm not sure about the name "inverse dissimilarity" (but, good idea); for clarification: is `compare(t1, t2).inverse_dissimilarity == compare(t2, t1).dissimilarity`? (No, right? What's the difference?)

I think you're right, we should call it something else (no idea what yet). 
`compare(t1, t2).inverse_dissimilarity == compare(t2, t1).dissimilarity` should be true (I checked experimentally with some large simulated tree sequences). It most likely could be proven because `inverse_dissmilarity` is simply looking for column max on our match matrix which is equivalent to the row max on the transpose of that matrix which is the match matrix for `compare(t2,t1)`. 

> And, maybe we (sigh) need another term, since we use "best match of n1 is n2" to mean something different than "best match of n2 is n1" (since "best match": T1 -> T2 is many-to-one).

Maybe we just reword "best match of n2 is n1" to "best match for n2 from the matching $\beta^{-1}\circ\beta$ from the fixed matching $beta: T_1\to T_2$. Not very pretty, but it is exact.

**Comment by petrelharp** [comment]
Thinking out loud here: let $M_{12}(n1,n2)$ be the span on which `n1` and `n2` agree; and let $\beta_{12}(n1)$ be such that
$$M_{12}(n1, \beta(n1)) = \max_{n2} M_{12}(n1, n2),$$
in other words (why can't I do \argmax here?):
$$\beta_{12}(n1) = argmax_{n2}\{M_{12}(n1, n2)\}.$$
Furthermore, let $\delta(n2) \in \beta_{12}^{-1}(n2)$ that maximizes $M$, i.e., 
$$\delta(n2) = argmax_{n1}\{M_{12}(n1, n2)\}.$$
Note that this is *a priori* not the same as
$$\beta_{21}(n2) = argmax_{n1}\{M_{21}(n2, n1)\},$$
since the maximum over all $n1$ might not be the same as the maximum over $\beta^{-1}(n2)$?

Indeed, suppose otherwise, i.e., $\beta_{21}(n2) = n1 \neq \delta(n2)$. Then (as long as there's no ties), we'd have that $M(n1,n2) > M(\delta(n2), n2)$. I think this is possible: suppose that we have $M(a1,b1) > M(a1, b2) > M(a2, b2)$, and that $\beta_{12}(a2) = b2$. Then $a1 \notin \beta_{12}^{-1}(b2)$, but $\beta_{21}(b2) = a1$.

**Comment by petrelharp** [comment]
So - correct me if I'm wrong, but I think our original proposal was to use $\delta$; and now have you implemented it using $\beta_{21}$? It's not clear to me yet which is the better way to go.

File: tscompare/methods.py [hunk 2]
```diff
@@ -425,9 +440,20 @@ def f(t):
...
+    best_n2_match = best_match_matrix.argmax(axis=0).A1
+    n2_match_mask = best_n1_match[best_n2_match] == np.arange(other.num_nodes)
+    best_match_n1_spans = shared_spans[np.arange(ts.num_nodes), best_n1_match].reshape(
+        -1
+    )
+    best_match_n2_spans = shared_spans[
+        best_n2_match, np.arange(other.num_nodes)
+    ].reshape(-1)[0, n2_match_mask]
+    total_match_n1_span = np.sum(best_match_n1_spans)
+    total_match_n2_span = np.sum(best_match_n2_spans)
```
Comment (hfr1tz3) [inline tscompare/methods.py:456]: We are currently using the $\delta$ approach. I took some time and I agree with you that there are cases where `compare(t,s).inverse_dissimilarity != compare(s,t).dissimilarity`. For instance, the match matrix
[3,1,2]
[0,2,1]
[4,5,1]
should fail.
Sometimes I wonder if such examples are constructible (like does there exist a well-defined pair of tree sequences with such a match matrix?)

Comment (hfr1tz3) [reply tscompare/methods.py:456]: I additionally think that it is better to go with the $\delta$ version because then it keeps are world dependent on a singular matching $\beta_{12}:T_1\to T_2$

Comment (petrelharp) [reply tscompare/methods.py:456]: > keeps are world
sorry my on-board imputation is failing on this one?

Comment (petrelharp) [reply tscompare/methods.py:456]: I think any example is constructable: for a given $M(n1,n2)$ suppose there are enough samples so that for each $(n1, n2)$ we can associate a unique sample set $S_{n1,n2}$; then define two tree sequences constructed as a sequence of intervals such that the $(n1,n2)$th interval is of length $M(n1,n2)$, and on that interval $n1$ is above $S_{n1,n2}$ on the first tree sequence and $n2$ is above (a polytomy of) the same set on the second tree sequence (and the rest of the samples are isolated); then $n1$ and $n2$ do not appear anywhere else along the genome.

**Comment by petrelharp** [comment]
Okay - I think I agree with you, the method using $\delta$ is the way to go. This is because we're thinking about the case where T2 is the truth and T1 is inferred, and we want to say something like "okay this node in T1 is standing in for that node in T2". If we do the $\beta_{21}$ method, then we'd be allowing a single node in T1 to be standing in for more than one node in T2, which seems like not what we're going for.

File: tscompare/methods.py [hunk 3]
```diff
@@ -425,9 +440,20 @@ def f(t):
...
     )
     # Between each pair of nodes, find the maximum shared span
-    best_match = best_match_matrix.argmax(axis=1).A1
-    best_match_spans = shared_spans[np.arange(len(best_match)), best_match].reshape(-1)
-    total_match_span = np.sum(best_match_spans)
+    # n1_match is the matching N1 -> N2 (for arf, dissimilarity)
+    # n2_match is finds the max match between nodes in N2 and their
+    # best match in N1 based on max-span (for tpr, inverse_dissimilarity)
+    best_n1_match = best_match_matrix.argmax(axis=1).A1
+    best_n2_match = best_match_matrix.argmax(axis=0).A1
```
Comment (petrelharp) [inline tscompare/methods.py:447]: I don't see how this can be the $\delta$ approach if it's based on this? (Since here the argmax is over all nodes?)

**Comment by petrelharp** [comment]
I don't think I agree this is currently doing the $\delta$ approach - see comment in the code above? Hm - in my initial comment above I omitted the $n1 \in \beta^{-1}(n2)$ from the subscript of the argmax in the definition of $\delta$ - I hope I didn't confuse things?

File: tscompare/methods.py (hunk 3 above)
Comment (hfr1tz3) [reply tscompare/methods.py:447]: The mask on line 448 is used to then check if n2s best match is n1 did n1 also match to n2 forcing it to be the $\delta$ method.

File: tscompare/methods.py (hunk 2 above)
Comment (hfr1tz3) [reply tscompare/methods.py:456]: > _edit:_ Oh is it "keeps our world"?
yes, and by 'our world' I just mean the construction.

> I think any example is constructable: ....
Oh this is clever. Thank you for the clairification

File: tscompare/methods.py (hunk 3 above)
Comment (hfr1tz3) [reply tscompare/methods.py:447]: I was going to say that the masking on line 448 should fix that problem, but now I don't think it does that...

Comment (petrelharp) [reply tscompare/methods.py:447]: Have a look at my (corrected) definition of $\delta$ again - that's now how it works?

Comment (hfr1tz3) [reply tscompare/methods.py:447]: Currently resolving. I found a nice way to fix it so that it is $\delta$.

**Comment by petrelharp** [comment]
That looks right, but with all the matrix-rearranging it's a bit hard for me to tell. Could you come up with a test that would have failed previously and now passes?

**Comment by hfr1tz3** [comment]
> That looks right, but with all the matrix-rearranging it's a bit hard for me to tell. Could you come up with a test that would have failed previously and now passes?

I'm having a hard time constructing the tree sequences which would have the shared span matrix:
[1 0]
[2 3]
but this matrix would be one which the above algorithm would satisfy now which wasn't possible previously.

**Comment by petrelharp** [comment]
How about this one?

**Comment by petrelharp** [comment]
How about this one?
Screenshot from 2025-01-20 20-45-21

**Comment by hfr1tz3** [comment]
Yes, this example mostly works. In fact, by coding up this example it made me realize that the `best_match_matrix` (line 436) is already filtered to only include the potential matches based on max shared span. So, we need to look specifically at pairs of tree sequences with multiple potential matches: ie. something with shared span matrix 
[3 3]
[1 2]
or something of the like. Luckily it was just a quick edit to your example.

File: tscompare/methods.py [hunk 4]
```diff
@@ -425,9 +440,25 @@ def f(t):
...
-    total_match_span = np.sum(best_match_spans)
+    # n1_match is the matching N1 -> N2 (for arf, dissimilarity)
+    # n2_match is finds the max match between nodes in N2 and their
+    # best match in N1 based on max-span (for tpr, inverse_dissimilarity)
+    best_n1_match = best_match_matrix.argmax(axis=1).A1
+    n2_match_matrix = scipy.sparse.lil_matrix((ts.num_nodes, other.num_nodes))
+    n2_match_matrix[np.arange(ts.num_nodes), best_n1_match] = best_match_matrix.tocsr()[
+        np.arange(ts.num_nodes), best_n1_match
+    ]
+    n2_match_matrix = n2_match_matrix.tocsr()
```
Comment (petrelharp) [inline tscompare/methods.py:451]: A maybe more efficient way to do this is something like this:
```
n2_match_matrix = best_match_matrix.tocsr()
bmm_row_ind = np.repeat(np.arange(best_match_matrix.shape[0]), repeats=np.diff(best_match_matrix.indptr))
n2_match_matrix[data] *= (best_match_matrix.indices == best_n1_match[bmm_row_ind])
```

File: tests/test_methods.py [hunk 5]
```diff
@@ -121,18 +121,30 @@ def f(t):
                         transform(ts.nodes_time[i]) - transform(other.nodes_time[j])
                     )
                     dissimilarity_matrix[i, j] = 1 / (1 + time_array[i, j])
-    best_match = np.argmax(dissimilarity_matrix, axis=1)
-    best_match_spans = np.zeros((ts.num_nodes,))
+    best_match_n1 = np.argmax(dissimilarity_matrix, axis=1)
+    n2_match_matrix = np.zeros((ts.num_nodes, other.num_nodes))
```
Comment (petrelharp) [inline tests/test_methods.py:125]: this should use shared spans not dissimiliarty (since that's times)

File: tscompare/methods.py [hunk 6]
```diff
@@ -438,14 +469,15 @@ def f(t):
...
     product = np.multiply((time_discrepancies**2), ts_node_spans)
     rmse = np.sqrt(np.sum(product) / total_span_ts)
     return ARFResult(
-        arf=1.0 - total_match_span / total_span_ts,
-        tpr=total_match_span / total_span_other,
-        dissimilarity=total_span_ts - total_match_span,
+        arf=1.0 - total_match_n1_span / total_span_ts,
+        tpr=total_match_n2_span / total_span_other,
+        dissimilarity=total_span_ts - total_match_n1_span,
+        inverse_dissimilarity=total_span_other - total_match_n2_span,
```
Comment (petrelharp) [inline tscompare/methods.py:480]: proposal to change these to to
```suggestion
        matched_span = (total_match_n1_span, total_match_n2_span)
```

**Comment by petrelharp** [comment]
I rewrote the naive method first (to be more naive), and turned up an issue in the implementation - I think - with `rmse`. Consider the case where a node has *no* best match (i.e., shared span is zero everywhere). Happily, there was a test for this! Should this node contributed to RMSE? I'd argue "no", it's only between nodes with a legit match. Previously, we were getting contributions to the RMSE from those nodes. So, now that test fails:
```
FAILED tests/test_methods.py::TestDissimilarity::test_with_no_match - AssertionError: 1.1702287212721978 != 0.46878005553759217
```
I guess we need to re-compute the naive expectation for that one.

**Comment by petrelharp** [comment]
Okay, done with the re-write! And, I might not have broke anything?!? @hfr1tz3 I hope you look at it carefully though.

And, I turned up another minor issue - we're accounting properly for missing data in `node_spans` but not in `shared_node_spans`. (This only affects sample nodes.) This is caught by the new `test_very_simple`.f

**Comment by petrelharp** [comment]
What happened to our CI, though?

**Comment by hfr1tz3** [comment]
> And, I turned up another minor issue - we're accounting properly for missing data in `node_spans` but not in `shared_node_spans`. (This only affects sample nodes.) This is caught by the new `test_very_simple`.f

Can you explain more what you mean by missing data? If this is a problem somewhere with the `shared_node_spans` code that's more of a question @nspope.
I also believe that even in our paper we always begin with tree sequences with the same sample size of sample sets so this example wouldn't occur.

**Comment by nspope** [comment]
'missing data' in that an edge connecting a sample node is removed (so the sample node is disconnected from the rest of the graph over part of the sequence). I don't think this'll happen for tsinfer'd tree sequences (or any other inference method I'm aware of), so maybe the immediate thing to do is detect it and error out?

**Comment by hfr1tz3** [comment]
Oh I see, I thought for a second the second Tree sequence was just 2 -> 0 with no sample 1.

**Comment by petrelharp** [comment]
I've dealt with the missing value thing (just declared "we don't deal with missing data"), and uncovered another potential issue: we aren't requiring samples to match themselves. (Suppose that two tree sequences are the same, but in one the sample IDs are shuffled.)

**Comment by nspope** [comment]
>  we aren't requiring samples to match themselves

That's just a documentation fix, I assume? As sample identity *is* the sample ID -- there's no way to ensure samples are identical between two distinct tree sequences in the absence of identifying metadata

**Comment by petrelharp** [comment]
Okay - done. I ran into those edge cases with some very simple weird examples. We ought to have another test or two that exercises these things - for instance, samples being parents to other nodes, etc.

**Comment by petrelharp** [comment]
> That's just a documentation fix, I assume? As sample identity is the sample ID -- there's no way to ensure samples are identical between two distinct tree sequences in the absence of identifying metadata

Here's the example: 
```
        # 1.00┊ 1 ┊         1.00┊ 0 ┊
        #     ┊ ┃ ┊   and       ┊ ┃ ┊
        # 0.00┊ 0 ┊         0.00┊ 1 ┊
        #     0   1             0   1
        # Note that if samples weren't required to match samples,
        # then 1 in the first matches 0 in the second (both are
        # ancestral to {0, 1})
```
The shared_span matrix here is
```
0 0
1 0
```
BUT we don't want to actually count that `1` as shared span. In other words, previously we would have said "oh yeah we're correctly representing the node `0` in the right-hand tree sequence, since we've got a node that's ancestral to {0, 1} in the left one"; however, we shouldn't have let that count, since that's maping one sample to another one. I deal with this by just zeroing out all entries `shared_spans[i,j]` where both `i` and `j` are samples but `i != j`.

Ah: and I added an error if the samples in the two tree sequences are not the same. So I need to document and test that. *Then* I'll be done.

**Comment by hfr1tz3** [comment]
Fixed some indexing errors in the algorithm and ran all the tests. 
Probably should clean up the test file a little more, but the bulk of the work is done

File: tscompare/methods.py [hunk 7]
```diff
@@ -430,9 +430,12 @@ def f(t):
...
     # but this is out of scope (people could detect this using
     # the shared spans matrix directly).
-    is_sample = np.full(ts.num_samples, False)
+    max_samples = max(ts.num_samples, other.num_samples)
+    is_sample = np.full(max(ts.num_nodes, other.num_nodes), False)
     is_sample[samples] = True
-    shared_spans.data[~np.logical_xor(is_sample[row_ind], is_sample[col_ind])] = 0.0
+    index_contains_sample = is_sample[row_ind]
+    index_not_equal = ~np.equal(row_ind, col_ind)
+    shared_spans.data[np.logical_and(is_sample[row_ind], index_not_equal)] = 0.0
```
Comment (petrelharp) [inline tscompare/methods.py:436]: Ah, whoops - good catch. But I think it's not quite there: right now there's no need to have `is_sample` the length of `other.num_nodes`, because it's only ever indexed by `row_ind`. Actually, this will produce an error if `ts.num_nodes != other.num_nodes` - both on indexing by `row_ind` and also since `np.equal( )` requires equal lengths.

(Also,  `index_contains_sample` isn't used?)

File: tscompare/methods.py [hunk 8]
```diff
@@ -32,12 +32,17 @@
 import tskit
 
 
-def node_spans(ts):
+def node_spans(ts, include_missing=False):
```
Comment (petrelharp) [inline tscompare/methods.py:35]: We weren't testing this option; I've added a test for that. However, we still don't have a test on a simple tree sequence where we know the right answer (instead of just comparing to the python version). Can you add a test of `node_spans` to those?

File: tscompare/methods.py (hunk 7 above)
Comment (petrelharp) [reply tscompare/methods.py:436]: So - looks like we need a test where the number of nodes in the two tree sequences isn't equal?

Comment (petrelharp) [reply tscompare/methods.py:436]: Hah, as you said, I'm totally wrong here. Never mind. =)

**Comment by petrelharp** [comment]
This looks good! There's one more thing, which is re-naming `dissimilarity` and `inverse_dissimilarity` (proposal was to a tuple, called `shared_spans` I think?). And, as you said, re-read the docs.

**Comment by hfr1tz3** [comment]
@petrelharp 
Ready for final review
🎉
